File: packages/python-sdk/strait/operations/jobs.py
"""Job management operations."""

from __future__ import annotations

from typing import Any

from strait.operations._base import AsyncBaseService, BaseService


class JobsService(BaseService):
    def list(self, *, query: dict[str, str] | None = None) -> dict[str, Any]:
        return self._request("GET", "/v1/jobs", query=query)

    def create(self, body: Any) -> dict[str, Any]:
        return self._request("POST", "/v1/jobs", body=body)

    def get(self, job_id: str) -> dict[str, Any]:
        return self._request("GET", "/v1/jobs/{jobID}", path_params={"jobID": job_id})

    def update(self, job_id: str, body: Any) -> dict[str, Any]:
        return self._request("PATCH", "/v1/jobs/{jobID}", path_params={"jobID": job_id}, body=body)

    def delete(self, job_id: str) -> dict[str, Any]:
        return self._request("DELETE", "/v1/jobs/{jobID}", path_params={"jobID": job_id})

    def clone(self, job_id: str, body: Any) -> dict[str, Any]:
        return self._request(
            "POST", "/v1/jobs/{jobID}/clone", path_params={"jobID": job_id}, body=body,
        )

    def get_health(self, job_id: str) -> dict[str, Any]:
        return self._request("GET", "/v1/jobs/{jobID}/health", path_params={"jobID": job_id})

    def trigger(self, job_id: str, body: Any) -> dict[str, Any]:
        return self._request(
            "POST", "/v1/jobs/{jobID}/trigger", path_params={"jobID": job_id}, body=body,
        )

    def bulk_trigger(self, job_id: str, body: Any) -> dict[str, Any]:
        return self._request(
            "POST", "/v1/jobs/{jobID}/trigger/bulk", path_params={"jobID": job_id}, body=body,
        )

    def list_versions(
        self, job_id: str, *, query: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        return self._request(
            "GET", "/v1/jobs/{jobID}/versions", path_params={"jobID": job_id}, query=query,
        )

    def get_version(self, job_id: str, version_id: str) -> dict[str, Any]:
        return self._request(
            "GET", "/v1/jobs/{jobID}/versions/{versionID}",
            path_params={"jobID": job_id, "versionID": version_id},
        )

    def list_dependencies(self, job_id: str) -> dict[str, Any]:
        return self._request(
            "GET", "/v1/jobs/{jobID}/dependencies", path_params={"jobID": job_id},
        )

    def create_dependency(self, job_id: str, body: Any) -> dict[str, Any]:
        return self._request(
            "POST", "/v1/jobs/{jobID}/dependencies", path_params={"jobID": job_id}, body=body,
        )

    def delete_dependency(self, job_id: str, dep_id: str) -> dict[str, Any]:
        return self._request(
            "DELETE", "/v1/jobs/{jobID}/dependencies/{depID}",
            path_params={"jobID": job_id, "depID": dep_id},
        )

    def batch(self, body: Any) -> dict[str, Any]:
        return self._request("POST", "/v1/jobs/batch", body=body)

    def batch_disable(self, body: Any) -> dict[str, Any]:
        return self._request("POST", "/v1/jobs/batch-disable", body=body)

    def batch_enable(self, body: Any) -> dict[str, Any]:
        return self._request("POST", "/v1/jobs/batch-enable", body=body)


class AsyncJobsService(AsyncBaseService):
    async def list(self, *, query: dict[str, str] | None = None) -> dict[str, Any]:
        return await self._request("GET", "/v1/jobs", query=query)

    async def create(self, body: Any) -> dict[str, Any]:
        return await self._request("POST", "/v1/jobs", body=body)

    async def get(self, job_id: str) -> dict[str, Any]:
        return await self._request("GET", "/v1/jobs/{jobID}", path_params={"jobID": job_id})

    async def update(self, job_id: str, body: Any) -> dict[str, Any]:
        return await self._request(
            "PATCH", "/v1/jobs/{jobID}", path_params={"jobID": job_id}, body=body,
        )

    async def delete(self, job_id: str) -> dict[str, Any]:
        return await self._request("DELETE", "/v1/jobs/{jobID}", path_params={"jobID": job_id})

    async def clone(self, job_id: str, body: Any) -> dict[str, Any]:
        return await self._request(
            "POST", "/v1/jobs/{jobID}/clone", path_params={"jobID": job_id}, body=body,
        )

    async def get_health(self, job_id: str) -> dict[str, Any]:
        return await self._request(
            "GET", "/v1/jobs/{jobID}/health", path_params={"jobID": job_id},
        )

    async def trigger(self, job_id: str, body: Any) -> dict[str, Any]:
        return await self._request(
            "POST", "/v1/jobs/{jobID}/trigger", path_params={"jobID": job_id}, body=body,
        )

    async def bulk_trigger(self, job_id: str, body: Any) -> dict[str, Any]:
        return await self._request(
            "POST", "/v1/jobs/{jobID}/trigger/bulk", path_params={"jobID": job_id}, body=body,
        )

    async def list_versions(
        self, job_id: str, *, query: dict[str, str] | None = None,
    ) -> dict[str, Any]:
        return await self._request(
            "GET", "/v1/jobs/{jobID}/versions", path_params={"jobID": job_id}, query=query,
        )

    async def get_version(self, job_id: str, version_id: str) -> dict[str, Any]:
        return await self._request(
            "GET", "/v1/jobs/{jobID}/versions/{versionID}",
            path_params={"jobID": job_id, "versionID": version_id},
        )

    async def list_dependencies(self, job_id: str) -> dict[str, Any]:
        return await self._request(
            "GET", "/v1/jobs/{jobID}/dependencies", path_params={"jobID": job_id},
        )

    async def create_dependency(self, job_id: str, body: Any) -> dict[str, Any]:
        return await self._request(
            "POST", "/v1/jobs/{jobID}/dependencies", path_params={"jobID": job_id}, body=body,
        )

    async def delete_dependency(self, job_id: str, dep_id: str) -> dict[str, Any]:
        return await self._request(
            "DELETE", "/v1/jobs/{jobID}/dependencies/{depID}",
            path_params={"jobID": job_id, "depID": dep_id},
        )

    async def batch(self, body: Any) -> dict[str, Any]:
        return await self._request("POST", "/v1/jobs/batch", body=body)

    async def batch_disable(self, body: Any) -> dict[str, Any]:
        return await self._request("POST", "/v1/jobs/batch-disable", body=body)

    async def batch_enable(self, body: Any) -> dict[str, Any]:
        return await self._request("POST", "/v1/jobs/batch-enable", body=body)

File: packages/ruby-sdk/lib/strait/operations/jobs.rb
# frozen_string_literal: true

module Strait
  module Operations
    # Service for job CRUD, triggering, versioning, and dependency management.
    class JobsService < BaseService
      # List all jobs.
      # GET /v1/jobs
      def list(query: nil)
        _request(:get, "/v1/jobs", query: query)
      end

      # Create a new job.
      # POST /v1/jobs
      def create(body)
        _request(:post, "/v1/jobs", body: body)
      end

      # Get a job by ID.
      # GET /v1/jobs/{jobID}
      def get(job_id)
        _request(:get, "/v1/jobs/{jobID}", path_params: { "jobID" => job_id })
      end

      # Update a job.
      # PATCH /v1/jobs/{jobID}
      def update(job_id, body)
        _request(:patch, "/v1/jobs/{jobID}", path_params: { "jobID" => job_id }, body: body)
      end

      # Delete a job.
      # DELETE /v1/jobs/{jobID}
      def delete(job_id)
        _request(:delete, "/v1/jobs/{jobID}", path_params: { "jobID" => job_id })
        nil
      end

      # Clone a job.
      # POST /v1/jobs/{jobID}/clone
      def clone(job_id, body)
        _request(:post, "/v1/jobs/{jobID}/clone", path_params: { "jobID" => job_id }, body: body)
      end

      # Get health status of a job.
      # GET /v1/jobs/{jobID}/health
      def get_health(job_id)
        _request(:get, "/v1/jobs/{jobID}/health", path_params: { "jobID" => job_id })
      end

      # Trigger a job run.
      # POST /v1/jobs/{jobID}/trigger
      def trigger(job_id, body)
        _request(:post, "/v1/jobs/{jobID}/trigger", path_params: { "jobID" => job_id }, body: body)
      end

      # Bulk-trigger a job.
      # POST /v1/jobs/{jobID}/trigger/bulk
      def bulk_trigger(job_id, body)
        _request(:post, "/v1/jobs/{jobID}/trigger/bulk", path_params: { "jobID" => job_id }, body: body)
      end

      # List versions of a job.
      # GET /v1/jobs/{jobID}/versions
      def list_versions(job_id, query: nil)
        _request(:get, "/v1/jobs/{jobID}/versions", path_params: { "jobID" => job_id }, query: query)
      end

      # Get a specific version of a job.
      # GET /v1/jobs/{jobID}/versions/{versionID}
      def get_version(job_id, version_id)
        _request(:get, "/v1/jobs/{jobID}/versions/{versionID}", path_params: { "jobID" => job_id, "versionID" => version_id })
      end

      # List dependencies for a job.
      # GET /v1/jobs/{jobID}/dependencies
      def list_dependencies(job_id)
        _request(:get, "/v1/jobs/{jobID}/dependencies", path_params: { "jobID" => job_id })
      end

      # Create a dependency for a job.
      # POST /v1/jobs/{jobID}/dependencies
      def create_dependency(job_id, body)
        _request(:post, "/v1/jobs/{jobID}/dependencies", path_params: { "jobID" => job_id }, body: body)
      end

      # Delete a dependency from a job.
      # DELETE /v1/jobs/{jobID}/dependencies/{depID}
      def delete_dependency(job_id, dep_id)
        _request(:delete, "/v1/jobs/{jobID}/dependencies/{depID}", path_params: { "jobID" => job_id, "depID" => dep_id })
        nil
      end

      # Batch-create jobs.
      # POST /v1/jobs/batch
      def batch(body)
        _request(:post, "/v1/jobs/batch", body: body)
      end

      # Batch-disable jobs.
      # POST /v1/jobs/batch-disable
      def batch_disable(body)
        _request(:post, "/v1/jobs/batch-disable", body: body)
      end

      # Batch-enable jobs.
      # POST /v1/jobs/batch-enable
      def batch_enable(body)
        _request(:post, "/v1/jobs/batch-enable", body: body)
      end
    end
  end
end
